### PR TITLE
Add QueueItemApi methods

### PR DESCRIPTION
**Pull request by martinda**
Adding a QueueItemApi to obtain the build number from the queue Id, so builds can be tracked from submission to completion. I am trying to implement code to ultimately arrive to [this answer](https://stackoverflow.com/questions/28311030/check-jenkins-job-status-after-triggering-a-build-remotely/48531874#48531874) for the jenkins-rest.

The queue item json looks like this:
```json
{
  "_class" : "hudson.model.Queue$LeftItem",
  "actions" : [
    {
      "_class" : "hudson.model.ParametersAction",
      "parameters" : [
        {
          "_class" : "hudson.model.StringParameterValue",
          "name" : "a",
          "value" : "2"
        }
      ]
    },
    {
      "_class" : "hudson.model.CauseAction",
      "causes" : [
        {
          "_class" : "hudson.model.Cause$UserIdCause",
          "shortDescription" : "Started by user martin",
          "userId" : "martin",
          "userName" : "martin"
        }
      ]
    }
  ],
  "blocked" : false,
  "buildable" : false,
  "id" : 14,
  "inQueueSince" : 1524061515284,
  "params" : "\na=2",
  "stuck" : false,
  "task" : {
    "_class" : "hudson.model.FreeStyleProject",
    "name" : "test",
    "url" : "http://localhost:8082/job/test/",
    "color" : "blue"
  },
  "url" : "queue/item/14/", 
  "why" : null,
  "cancelled" : false,
  "executable" : {
    "_class" : "hudson.model.FreeStyleBuild",
    "number" : 3,
    "url" : "http://localhost:8082/job/test/3/"
  }
}
```
The "executable" entry tells us the build number. If the submission is canceled before the build is launched, it looks like this:
```json
...
  "cancelled" : true,
  "executable" : null
}
```

### CODE REVIEW

File: src/main/java/com/cdancy/jenkins/rest/features/QueueItemApi.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import com.cdancy.jenkins.rest.domain.queue.QueueItem;
+import com.cdancy.jenkins.rest.filters.JenkinsAuthenticationFilter;
+
+@RequestFilters(JenkinsAuthenticationFilter.class)
+@Consumes(MediaType.APPLICATION_JSON)
+@Path("/queue/item/{queueId}/api/json")
+public interface QueueItemApi {
+
+   @Named("queueItem:queueItem")
+   @SelectJson("")
```
Comment (martinda) [inline src/main/java/com/cdancy/jenkins/rest/features/QueueItemApi.java]: @cdancy, What do I set the `@SelectJson` argument value to, to return the queue item itself with all its fields? The json does not have a "top" name for individual items.

Comment (cdancy) [reply src/main/java/com/cdancy/jenkins/rest/features/QueueItemApi.java]: You can remove it altogether so long as the returnes Object matches the main returned from the server.

**Comment by cdancy** [comment]
If push back on creating this new API and instead just add this endpoint to the existing QueueApi class. It will have to be slightly massaged but I can help with that if need be.

**Comment by martinda** [comment]
Oh, I see. Ok, I can try to move it inside the existing queue api. But one word of caution: once the build starts, the queue no longer lists the queue item, it rather seems to "float" at the top of jenkins under `/queue/item`.

File: src/main/java/com/cdancy/jenkins/rest/features/QueueItemApi.java (hunk 1 above)
Comment (martinda) [reply src/main/java/com/cdancy/jenkins/rest/features/QueueItemApi.java]: I tried removing it, but it is the same as setting to an empty string: the queue item is a null. I am stuck but I am not giving up just yet.

Comment (martinda) [reply src/main/java/com/cdancy/jenkins/rest/features/QueueItemApi.java]: I think my groovy grapes was not re-fetching from my latest local maven repo.

**Comment by martinda** [comment]
Now working on the integration tests. Fun stuff.

**Comment by cdancy** [comment]
A couple minor comments to address, as well as an integration test, and things should be good to go. Code looks clean and thanks for the handful of mockTests you knocked out.

> Now working on the integration tests. Fun stuff.

I agree. It's the reason I used the jclouds library in the backend as it makes writing this stuff a breeze and it sort of all just flows together IMO.

File: src/main/java/com/cdancy/jenkins/rest/features/QueueApi.java [hunk 2]
```diff
@@ -21,30 +21,67 @@
...
+    // < Server: Jetty(9.4.z-SNAPSHOT)
+    // <
+    // * Connection #0 to host localhost left intact
+    @Named("queue:cancel")
+    @Path("/queue/cancelItem")
+    //@Fallback(Fallbacks.NullOnNotFoundOr404.class)
+    //@Fallback(JenkinsFallbacks.RequestStatusOnError.class)
+    @ResponseParser(RequestStatusParser.class)
+    @POST
+    RequestStatus cancel(@FormParam("id") Integer queueId);
```
Comment (martinda) [inline src/main/java/com/cdancy/jenkins/rest/features/QueueApi.java]: @cdancy I am really puzzled at how to make this POST work. The integration test that uses it keeps failing with a RequestStatus value of false, as if the cancel had not worked, but it does seem to work when I read the log.

File: src/test/java/com/cdancy/jenkins/rest/features/QueueApiLiveTest.java [hunk 3]
```diff
@@ -57,6 +59,87 @@ public void testGetQueue() {
...
+
+    @Test
+    public void testGetCancelledQueueItem() throws InterruptedException {
+        Integer job1 = api.jobsApi().build("QueueTest");
+        assertNotNull(job1);
+        Integer job2 = api.jobsApi().build("QueueTest");
+        assertNotNull(job2);
+
+        RequestStatus success = api().cancel(job2);
+        assertTrue(success.value());
```
Comment (martinda) [inline src/test/java/com/cdancy/jenkins/rest/features/QueueApiLiveTest.java]: @cdancy. This assert never succeeds. The log says there is a redirection (302) and I am not sure how to deal with that...
```
    21:59:15.061 [Test worker] DEBUG org.jclouds.rest.internal.InvokeHttpMethod - >> invoking queue:cancel
    21:59:15.061 [Test worker] DEBUG org.jclouds.http.internal.JavaUrlHttpCommandExecutorService - Sending request 830640781: POST http://localhost:8082/queue/cancelItem HTTP/1.1
    21:59:15.062 [Test worker] DEBUG jclouds.wire - >> "id=257"
    21:59:15.062 [Test worker] DEBUG jclouds.headers - >> POST http://localhost:8082/queue/cancelItem HTTP/1.1
    21:59:15.062 [Test worker] DEBUG jclouds.headers - >> Accept: application/json
    21:59:15.062 [Test worker] DEBUG jclouds.headers - >> Authorization: Basic bWFydGluOm1hcnRpbg==
    21:59:15.062 [Test worker] DEBUG jclouds.headers - >> Jenkins-Crumb: 53842b63c40e494a74d261a4773339ef
    21:59:15.062 [Test worker] DEBUG jclouds.headers - >> Content-Type: application/x-www-form-urlencoded
    21:59:15.062 [Test worker] DEBUG jclouds.headers - >> Content-Length: 6
    21:59:15.157 [Test worker] DEBUG org.jclouds.http.internal.JavaUrlHttpCommandExecutorService - Receiving response 830640781: HTTP/1.1 302 Found
    21:59:15.157 [Test worker] DEBUG jclouds.headers - << HTTP/1.1 302 Found
    21:59:15.157 [Test worker] DEBUG jclouds.headers - << Server: Jetty(9.4.z-SNAPSHOT)
    21:59:15.157 [Test worker] DEBUG jclouds.headers - << X-Content-Type-Options: nosniff
    21:59:15.157 [Test worker] DEBUG jclouds.headers - << Date: Thu, 19 Apr 2018 01:59:15 GMT
    21:59:15.157 [Test worker] DEBUG jclouds.headers - << Location: http://localhost:8082/queue/
    21:59:15.157 [Test worker] DEBUG jclouds.headers - << Content-Type: application/unknown
    21:59:15.157 [Test worker] DEBUG jclouds.headers - << Content-Length: 0
    21:59:15.158 [Test worker] DEBUG org.jclouds.http.internal.JavaUrlHttpCommandExecutorService - Sending request -1390855584: POST http://localhost:8082/queue/ HTTP/1.1
    21:59:15.158 [Test worker] DEBUG jclouds.wire - >> "id=257"
    21:59:15.158 [Test worker] DEBUG jclouds.headers - >> POST http://localhost:8082/queue/ HTTP/1.1
    21:59:15.158 [Test worker] DEBUG jclouds.headers - >> Accept: application/json
    21:59:15.158 [Test worker] DEBUG jclouds.headers - >> Authorization: Basic bWFydGluOm1hcnRpbg==
    21:59:15.158 [Test worker] DEBUG jclouds.headers - >> Jenkins-Crumb: 53842b63c40e494a74d261a4773339ef
    21:59:15.159 [Test worker] DEBUG jclouds.headers - >> Content-Type: application/x-www-form-urlencoded
    21:59:15.159 [Test worker] DEBUG jclouds.headers - >> Content-Length: 6
    21:59:15.252 [Test worker] DEBUG org.jclouds.http.internal.JavaUrlHttpCommandExecutorService - Receiving response -1390855584: HTTP/1.1 404 Not Found
    21:59:15.253 [Test worker] DEBUG jclouds.headers - << HTTP/1.1 404 Not Found
    21:59:15.253 [Test worker] DEBUG jclouds.headers - << Server: Jetty(9.4.z-SNAPSHOT)
    21:59:15.253 [Test worker] DEBUG jclouds.headers - << X-Content-Type-Options: nosniff
    21:59:15.253 [Test worker] DEBUG jclouds.headers - << Date: Thu, 19 Apr 2018 01:59:15 GMT
```

Comment (martinda) [reply src/test/java/com/cdancy/jenkins/rest/features/QueueApiLiveTest.java]: Found that the redirect to a non-existent page is actually issue [JENKINS-21311](https://issues.jenkins-ci.org/browse/JENKINS-21311)

**Comment by martinda** [comment]
@cdancy, I have completed the bulk of the work, and provided some tests. Tracking builds from submission to the queue, through to build completion is possible with this addition. I can add a groovy example to the readme file if that's where it belongs.

**Comment by cdancy** [comment]
@martinda massaged your branch a bit to get things working with the new `RequestStatus` model that I ported over from the `bitbucket-rest` project. I'm going to be converting all endpoints where applicable to this model so that nothing will ever fail or thrown an exception but instead be left up to the user to check for the existence of the `errors()` object and move forward appropriately.

Branch works as-is. Have a look over things and once I have your sign-off we can merge and release.

**Comment by martinda** [comment]
@cdancy Thank you. All the tests pass for me as well. I would have liked to clean up my commits with rebase, but `gradlew.bat` keeps getting in the way with CRLF translations. So if you just merge it as is I am good with that.

**Comment by cdancy** [comment]
No worries about the squashing. I'll knock out another release today.

**Pull request merged**